components/builder-depot/src/backend/s3.rs

```diff
@@ -0,0 +1,322 @@
+    //                                        if object.key.unwrap() == key {
+    //                                            return Ok(true);
+    //                                        }
+    //                                    }
+    //                                    return Ok(false);
+    //                                }
+    //                                None => return Ok(false),
+    //                            },
+    //                            Err(e) => {
+    //                                debug!("{:?}", e);
+    //                                return Err(Error::ObjectError(e));
+    //                            }
+    //                        }
+    //                    }
+    //                }
+    //                return Ok(false);
+    //            }
+    //            None => Ok(false),
+    //        },
+    //        Err(e) => {
+    //            debug!("{:?}", e);
+    //            Err(Error::BadBucket(e))
+    //        }
+    //    }
+    //}
+
+    // This function checks whether or not the
+    // configured bucket exists in the configured
+    // backend.
+    fn exists(&self) -> Result<bool> {
+        let artifactbucket = self.bucket.to_owned();
+        match self.client.list_buckets().sync() {
+            Ok(bucket_list) => match bucket_list.buckets {
+                Some(buckets) => {
+                    for bucket in buckets {
+                        if bucket.name.unwrap() == artifactbucket {
+                            return Ok(true);
+                        } else {
+                            return Ok(false);
+                        }
+                    }
+                    return Ok(false);
+                }
+                None => Ok(false),
+            },
+            Err(e) => {
+                debug!("{:?}", e);
+                return Err(Error::ListBuckets(e));
+            }
+        }
+    }
+
+    pub fn create_bucket(&self) -> Result<()> {
+        let mut request = CreateBucketRequest::default();
+        request.bucket = self.bucket.clone();
+
+        if self.exists().unwrap() {
+            Ok(())
+        } else {
+            match self.client.create_bucket(&request).sync() {
+                Ok(_response) => Ok(()),
+                Err(e) => {
+                    debug!("{:?}", e);
+                    Err(Error::CreateBucketError(e))
+                }
+            }
+        }
+    }
+
+    pub fn upload(
+        &self,
+        hart: &PathBuf,
+        ident: &PackageIdent,
+        target: &PackageTarget,
+    ) -> Result<()> {
+        let key = self.s3_key(ident, target).to_string_lossy().into_owned();
+        let file = File::open(hart)?;
+        let size = file.metadata().unwrap().len() as usize;
+        let min_limit: usize = 49152 * 1024;
+        if size < min_limit {
+            self.single_upload(key, file)
+        } else {
+            self.multipart_upload(key, file)
+        }
```

Comment:
One other thing to consider is adding timing data for upload and downloads, similiar to how the graph operations are instrumented - https://github.com/habitat-sh/builder/blob/master/components/builder-jobsrv/src/server/mod.rs#L153

Also, it would be good to add metrics so we can also track them on a dashboard like we do github calls.

Reply:
Absolutely. Do we want metrics on upload requests, successful uploads, and failed uploads?